code/modules/mob/living/carbon/xenomorph/hive_datum.dm

```diff
@@ -0,0 +1,422 @@
+// *********** Removing xenos
+// ***************************************
+/datum/hive_status/proc/remove_xeno(mob/living/carbon/Xenomorph/X) // should only be called by remove_from_hive
+	if(X.stat == DEAD)
+		if(!dead_xenos.Remove(X))
+			stack_trace("failed to remove a dead xeno from hive status dead list, nothing was removed!?")
+			return FALSE
+	else
+		remove_from_lists(X)
+
+	post_removal(X)
+	return TRUE
+
+// helper function
+/datum/hive_status/proc/remove_from_lists(mob/living/carbon/Xenomorph/X)
+	// Remove() returns 1 if it removes an element from a list
+	if(!xenos_by_tier[X.tier].Remove(X))
+		stack_trace("failed to remove a xeno from hive status tier list, nothing was removed!?")
+		return FALSE
+
+	if(!xenos_by_upgrade[X.upgrade].Remove(X))
+		stack_trace("trying to remove a xeno from hivestatus upgrade list, nothing was removed!?")
+		return FALSE
+
+	if(!xenos_by_typepath[X.caste_base_type])
+		stack_trace("trying to remove an invalid typepath from hivestatus list")
+		return FALSE
+
+	if(!xenos_by_typepath[X.caste_base_type].Remove(X))
+		stack_trace("failed to remove a xeno from hive status typepath list, nothing was removed!?")
+		return FALSE
+	
+	remove_leader(X)
+
+	return TRUE
+
+/mob/living/carbon/Xenomorph/proc/remove_from_hive()
+	if(!istype(hive))
+		CRASH("tried to remove a xeno from a hive that didnt have a hive to be removed from")
+
+	if(!hive.remove_xeno(src))
+		CRASH("failed to remove xeno from a hive")
+
+	if(queen_chosen_lead || src in hive.xeno_leader_list)
+		hive.remove_leader(src)
+
+	hive = null
+	hivenumber = XENO_HIVE_NONE // failsafe value
+
+/mob/living/carbon/Xenomorph/Queen/remove_from_hive() // override to ensure proper queen/hive behaviour
+	var/datum/hive_status/HS = hive
+	if(HS.living_xeno_queen == src)
+		HS.living_xeno_queen = null
+	. = ..()
+	HS.update_queen()
+
+// ***************************************
+// *********** Xeno leaders
+// ***************************************
+/datum/hive_status/proc/add_leader(mob/living/carbon/Xenomorph/X)
+	xeno_leader_list += X
+	X.queen_chosen_lead = TRUE
+
+/datum/hive_status/proc/remove_leader(mob/living/carbon/Xenomorph/X)
+	xeno_leader_list -= X
+	X.queen_chosen_lead = FALSE
+
+/datum/hive_status/proc/update_leader_pheromones() // helper function to easily trigger an update of leader pheromones
+	for(var/i in xeno_leader_list)
+		var/mob/living/carbon/Xenomorph/X = i
+		X.handle_xeno_leader_pheromones(living_xeno_queen)
+
+// ***************************************
+// *********** Xeno upgrades
+// ***************************************
+/datum/hive_status/proc/upgrade_xeno(mob/living/carbon/Xenomorph/X, oldlevel, newlevel) // called by Xenomorph/proc/upgrade_xeno()
+	xenos_by_upgrade[oldlevel] -= X
+	xenos_by_upgrade[newlevel] += X
+
+// ***************************************
+// *********** Xeno death
+// ***************************************
+/datum/hive_status/proc/on_xeno_death(mob/living/carbon/Xenomorph/X)

```

Comment:
Wouldn't this work better on a mob proc? They are more efficient even.

Reply:
its for hive datum behaviour that needs to happen when a xeno dies

Reply:
Agreed with spooky here